src/django_components/component.py

```diff
@@ -254,33 +258,76 @@ def get_template(self, context: Mapping) -> Template:
             f"Note: this attribute is not required if you are overriding the class's `get_template*()` methods."
         )
 
+    def render_from_input(self, context: Context, args: Union[List, Tuple], kwargs: Dict) -> str:
+        component_context: dict = self.get_context_data(*args, **kwargs)
+
+        with context.update(component_context):
+            rendered_component = self.render(context, context_data=component_context)
+
+        if is_dependency_middleware_active():
+            output = RENDERED_COMMENT_TEMPLATE.format(name=self.registered_name) + rendered_component
+        else:
+            output = rendered_component
+
+        return output
+
     def render(
         self,
-        context_data: Union[Dict[str, Any], Context],
+        context: Union[Dict[str, Any], Context],
         slots_data: Optional[Dict[SlotName, str]] = None,
         escape_slots_content: bool = True,
+        context_data: Optional[Dict[str, Any]] = None,
     ) -> str:
         # NOTE: This if/else is important to avoid nested Contexts,
         # See https://github.com/EmilStenstrom/django-components/issues/414
-        context = context_data if isinstance(context_data, Context) else Context(context_data)
-        prepare_context(context, component_id=self.component_id, outer_context=self.outer_context or Context())
+        context = context if isinstance(context, Context) else Context(context)
+        prepare_context(context, self.component_id)
         template = self.get_template(context)
 
-        # Associate the slots with this component for this context
-        # This allows us to look up component-specific slot fills.
-        def on_node(node: Node) -> None:
-            if isinstance(node, SlotNode):
-                trace_msg("ASSOC", "SLOT", node.name, node.node_id, component_id=self.component_id)
-                set_slot_component_association(context, node.node_id, self.component_id)
+        # Support passing slots explicitly to `render` method
+        if slots_data:
+            fill_content = self._fills_from_slots_data(slots_data, escape_slots_content)
```

Comment:
Renamed `_fill_slots` to `_fills_from_slots_data`, and also changed it so it does not change the value of `self.fill_slots`. So calling `Component.render` DOES NOT change the state of `Component` instance.